pennylane/measurements/shots.py

```diff
@@ -0,0 +1,176 @@
+    return isinstance(s, tuple) and len(s) == 2 and valid_int(s[0]) and valid_int(s[1])
+
+
+class Shots:
+    """
+    A data class that stores shot information.
+
+    Args:
+        shots (Union[None, int, Sequence[int, Tuple[int, int]]]): Raw shot information
+
+    Defining shots enables users to specify circuit executions, and the Shots class standardizes
+    the internal representation of shots. There are three ways to specify shot values:
+
+    * The value ``None``
+    * A positive integer
+    * A sequence consisting of either positive integers or a tuple-pair of positive integers of the form ``(shots, copies)``
+
+    The tuple-pair of the form ``(shots, copies)`` is represented internally by a namedtuple called
+    :class:`~ShotCopies`. The first value is the number of shots to execute, and the second value is the
+    number of times to repeat a circuit with that number of shots.
+
+    The ``Shots`` class exposes two properties:
+
+    * ``total_shots``, the total number of shots to be executed
+    * ``shot_vector``, the tuple of :class:`~ShotCopies` to be executed
+
+    Instances of this class are static. If an instance is passed to the constructor, that same
+    instance is returned. If an instance is constructed with a ``None`` value, ``total_shots``
+    will be ``None``.  This indicates analytic execution.
+
+    **Examples**
+
+    Example constructing a Shots instance with ``None``:
+
+    >>> shots = Shots(None)
+    >>> shots.total_shots, shots.shot_vector
+    (None, ())
+
+    Example constructing a Shots instance with an int:
+
+    >>> shots = Shots(100)
+    >>> shots.total_shots, shots.shot_vector
+    (100, (ShotCopies(shots=100, copies=1),))
+
+    Example constructing a Shots instance with another instance:
+
+    >>> shots = Shots(100)
+    >>> Shots(shots) is shots
+    True
+
+    Example constructing a Shots instance with a sequence of ints:
+
+    >>> shots = Shots([100, 200])
+    >>> shots.total_shots, shots.shot_vector
+    (300, (ShotCopies(shots=100, copies=1), ShotCopies(shots=200, copies=1)))
+
+    Example constructing a Shots instance with a sequence of tuple-pairs:
+
+    >>> shots = Shots(((100, 3), (200, 4),))
+    >>> shots.total_shots, shots.shot_vector
+    (1100, (ShotCopies(shots=100, copies=3), ShotCopies(shots=200, copies=4)))
+
+    Example constructing a Shots instance with a sequence of both ints and tuple-pairs.
+    Note that the first stand-alone ``100`` gets absorbed into the subsequent tuple because the
+    shot value matches:
+
+    >>> shots = Shots((10, 100, (100, 3), (200, 4),))
+    >>> shots.total_shots, shots.shot_vector
+    (1210, (ShotCopies(shots=10, copies=1), ShotCopies(shots=100, copies=4), ShotCopies(shots=200, copies=4)))
+
+    One should also note that specifying a single tuple of length 2 is considered two different
+    shot values, and *not* a tuple-pair representing shots and copies to avoid special behaviour
+    depending on the iterable type:
```

Comment:
Thanks for adding this note! I think including something along the lines of "To pass a single ``(shots, copies)`` pair, it must be wrapped within another sequence" would be a good idea.

Reply:
I just gave an extra example to illustrate the different meanings ✅